specs/spec.md

```diff
@@ -0,0 +1,319 @@
+    - [PING](#ping)
+6. [Implementations](#implementations)
+
+## Abstract
+
+Hobbits is a modular wire protocol which allows implementers to experiment with various application logic over a practical network in an agnostic and expedient manner.
+
+Hobbits wire protocol was developed so that Eth2.0 clients could begin network-level testing without requiring libp2p.
+
+## Flow
+
+We layout a network where the nodes are peered in the following manner:
+
+![nodes](./nodes.png)
+
+The following flow dictates how blocks are shared amongst peers:
+
+![sequence diagram](./diagram.png)
+
+1. Node 1 receives a new block `0xCAB` and gossips `hash(0xCAB)` to Node 2 and Node 3
+2. Node 2 requests the block bodies for block `0xCAB` from Node 1 in an RPC Command query
+3. Node 1 responds to Node 2 with the block bodies for block `0xCAB` in an RPC Command response
+4. Node 2 receives a new block `0xCAB` and gossips it to Node 1 and Node 3
+
+## Wire Protocol
+
+### Messages
+
+Messages are sent as `utf-8 byte encoded arrays`. The `message` format looks as follows:
+
+```
+EWP <version> <protocol> <header-length> <body-length>
+<header><body>
+```
+
+**NOTE:** Message`body`s along with `header` fields are BSON encoded.
+
+#### Fields
+
+Every hobbit message contains the following fields: 
+
+| Field | Definition | Validity |
+|:------:|----------|:----:|
+| `version` | Defines the EWP version number e.g. `0.2`. | `(\d+\.)(\d+)` |
+| `protocol` | Defines the [protocol](#protocols). | `(RPC\|GOSSIP\|PING)` |
+| `header` | Defines the header | payload |
+| `body` | Defines the body | payload |
+
+## Transport
+
+<!-- @TODO DISCUSS TCP AND SHIT -->
+
+## Protocols
+
+Hobbits defines 3 protocols that dictate how messages are interpreted and responded to.
+
+### RPC
+
+<!-- @TODO WRITE SOME SHIT -->
+
+#### Envelope
+
+```python
+{ 
+  'method_id': 'uint16' ## byte representing the method
+  'id': 'uint64' ## id of the request
+  'body': 'bytes' ## body of the request itself
+}
+```
+
+The `body` field contains an [SSZ](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md) encoded byte array. 
+
+##### Example
+
+Below you will find an example `RPC` request.
+
+```
+EWP 0.2 RPC 0 65
+{
+  "method_id": 0x01,
+  "id": 1,
+  "body": {
+    "bodies": "0x2a00000000000000"
+  }
+}
+```
+#### RPC Methods
+
+There are two types of RPC methods: one that requests data and one that responds with data.
+
+* Peers may request blocks and headers from other peers, possibly in response to a `GOSSIP` message. 
+* Peers may request blocks repeatedly from the same peers.
+* Other peers may respond on a best effort basis with header and block data.
+* There is no SLA for responding. 
+
+##### Handshake
+
+###### `0x00` HELLO
+
+Upon discovering each other, nodes may exchange `HELLO` messages.
+
+Nodes may send `HELLO` to other peers when they exchange messages for the first time or when their state changes to let them know new blocks are available.
```

Comment:
Don't we already let peers know that new blocks are available with GOSSIP anyway?

Reply:
Oh is this to let them know about the new head? rather than the new gossipped block?

Reply:
It's when you just come up the first time. You don't know anything about the state of the network and you need to sync up.